### PR TITLE
Replace well-known LaTeX commands with their unicode equivalent

### PR DESCRIPTION
**Pull request by timjb**
in text expression inputs. Under the hood, this uses the [unicodeit](svenkreiss/unicodeit) library, which powers [unicodeit.net](https://www.unicodeit.net/).

https://user-images.githubusercontent.com/124073/103143693-0ac0dc80-471c-11eb-84e1-551371c8aaea.mp4

TODO:

* [x] Provide a suggestion menu.
* [x] Perform the replacement only after the user confirms using the enter key or by leaving the input field. This solves the problem that otherwise it is not possible to input λ by typing `\lambda`, since the prefix `\l` is replaced by ł.

### CODE REVIEW

**Comment by SReichelt** [comment]
(I'm typing this after seeing just the first commit, so it might be somewhat obsolete.)

This is quite nice already. Thank you. I'd like to make some minor changes tomorrow; I could just push them to your branch if that's OK.

Regarding the suggestions menu, I could actually imagine two different menus: After typing `\`, there could be some auto-completion like in a text editor (as you probably had in mind). BTW, in Lean/VSCode, you can type e.g. `\al` followed by a space to get an alpha, which is quite nice. But in addition there could also be a menu _before_ typing `\`, with rows like "greek", "calligraphic", "symbol", etc.

Regarding `mathcal` and friends (especially bold and fraktur), I'm in favor of sending a patch upstream. However, to make it really convenient, I think we should provide single-key shortcuts like e.g. `/` for calligraphic. We could display those keys in the menu to make them discoverable.
Incidentally, in `unicode.ts` there is already some code for conversion in the reverse direction (because the font doesn't actually support the necessary characters).

Regarding the second point, it would be especially nice if input was converted as soon as it becomes unambiguous. Not sure if that can be achieved with the library.

**Comment by timjb** [comment]
@SReichelt I've finished the implementation of the autocomplete menu. The behavior is as follows:

* The input is converted as soon as it becomes unambiguous.
* Complete commands that are the prefix of another command (e.g. `\l` is a prefix of `\lambda`) are converted when the user presses enter or leaves the input field.

I would like to get this merged as soon as possible. There sure are a lot of potential improvements, but I think that it would be better to discuss and implement them separately. One improvement that I would particularly like to see is the ability to select a suggestion in the menu using the keyboard.

The latest version only uses the translation table from `unicodeit`, not its (quite primitive/hacky) `replace` function. I believe that if we want to make the LaTeX input really great, we will need to implement it ourselves, taking `unicodeit`'s table as a starting point.

**Review by SReichelt — CHANGES_REQUESTED** [review]
Sorry for the delay; I'm not at home right now.
I've fixed the two things that were broken. We could merge the current state and then continue from there, if you wish.
I would suggest a few changes, in decreasing order of importance:
- Right now, if you type e.g. `\alpha` blindly, you end up with "αha". The user cannot easily guess how many characters to type. So the conversion should either be delayed until the user types Enter or leaves the input, or additional input should be suppressed after a successful conversion.
- The initial menu after typing `\` shows some random entries. Maybe sort alphabetically?
- I'm currently not able to trigger conversion manually if it is ambiguous, even though there is code for that. But maybe I haven't understood it properly. (Once it works, we could highlight the entry that will be inserted in the menu.)
- After typing `\`, the font changes, but the text is still shown in the default font in other locations. I've pushed a possible solution in a second commit, but it is definitely not the only solution.
- I think we should make it a bit more generic so it also works in the "Symbol/Text" input within the notation dialog. I might take care of this myself.

Thanks again.

File: src/client/components/Expression.tsx [hunk 1]
```diff
@@ -982,113 +942,163 @@ class Expression extends React.Component<ExpressionProps, ExpressionState> {
...
+      className += ' definition';
+    }
+    let hasMenu = this.props.interactionHandler !== undefined && onMenuOpened !== undefined;
+    this.hasMenu = hasMenu;
+    let hasVisibleMenu = hasMenu && (alwaysShowMenu || this.isDirectlyHighlighted());
+    if (hasMenu || expression !== undefined) {
+      return this.wrapRenderedExpressionWithMenuStuff(result, expression, semanticLinks, onMenuOpened, hasMenu, hasVisibleMenu, className, innerClassName);
     } else {
-      className += innerClassName;
+      className += ' ' + innerClassName;
```
[the remark this answers is not in the document]
Comment (timjb) [reply src/client/components/Expression.tsx]: A lot of the `className`-related code could be simplified by using a library like [clsx](https://www.npmjs.com/package/clsx).

File: src/client/components/Expression.tsx [hunk 2]
```diff
@@ -1246,7 +1251,8 @@ class Expression extends React.Component<ExpressionProps, ExpressionState> {
   }
 
   private isDirectlyHighlighted(): boolean {
-    return (this.state.hovered && this.hoveredChildren.indexOf(this) >= 0) || this.permanentlyHighlighted || this.state.openMenu !== undefined;
+    return false;
```
[the remark this answers is not in the document]
Comment (timjb) [reply src/client/components/Expression.tsx]: No intentional change, I was trying things out and forgot to change them back.

File: src/client/components/Expression.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SReichelt) [reply src/client/components/Expression.tsx]: Good point. I've added a TODO.

**Comment by timjb** [comment]
@SReichelt I believe that all your points except the last one (and maybe the third one, which I don't quite understand) are now addressed. Could you please give this PR another review? :-)

**Comment by SReichelt** [comment]
Nice. :-)
Regarding the third point: I had seen some code that reacted to the Enter key, but I didn't know how to trigger it. I think now I understand: It was for the case when the user had typed `\l` which is both a prefix and an exact match. I was expecting to be able to use it in ambiguous cases as well, like in an IDE. In 2486f10, I've changed the code to support this. If you are OK with the change, I'll merge the PR tomorrow. (As a user, I'd kind of expect the up/down arrow keys to work now, but that's a separate task.)

**Review by SReichelt — APPROVED** [review]

**Comment by timjb** [comment]
I don't have a strong opinion on how exactly the autocomplete should behave, so I'm okay with merging this.

I am not happy about the fact that we have wasted effort changing the behavior multiple times. Before starting with the implementation of the next feature, I will initiate a discussion so we can come to an agreement about its details.

**Comment by SReichelt** [comment]
Looking back at the history, I hadn't properly considered these details myself. I didn't remember writing "it would be especially nice if input was converted as soon as it becomes unambiguous", and in hindsight that was clearly a mistake. I like the current behavior, though.
I have adapted the tutorial to showcase the feature, and merged the branch.

**Pull request merged**